octue/mixins/taggable.py

```diff
@@ -51,51 +61,75 @@ def _clean(tags):
             )
 
         # Strip leading and trailing whitespace, ensure they match the regex
-        cleaned_tags = []
+        cleaned_tags = set()
         for tag in tags:
             cleaned_tag = tag.strip()
             if not re.match(TAG_PATTERN, cleaned_tag):
                 raise InvalidTagException(
                     f"Invalid tag '{cleaned_tag}'. Tags must contain only characters 'a-z', '0-9', ':' and '-'. They must not start with '-' or ':'."
                 )
-            cleaned_tags.append(cleaned_tag)
+            cleaned_tags.add(cleaned_tag)
 
         return cleaned_tags
 
-    def __str__(self):
-        """ Represents tags as a space delimited string
-        """
-        return self.serialise()
+    def _yield_subtags(self, tags=None):
+        """ Yield the colon-separated subtags of a tag as strings, including the main tag. """
+        for tag in tags or self._tags:
+            yield from tag.split(":")
 
-    def serialise(self):
-        """ Serialises tags as a space delimited string, NOT as a list. Strips end whitespace.
+    def add_tags(self, *args):
+        """ Adds one or more new tag strings to the object tags. New tags will be cleaned and validated.
         """
-        return " ".join(self._tags).strip()
+        self._tags |= self._clean(args)
 
-    def startswith(self, value):
-        """ Implement a startswith method that returns true if any of the tags starts with value
+    def has_tag(self, tag):
+        """ Returns true if any of the tags exactly matches value, allowing test like `if 'a' in TagGroup('a b')`
         """
-        return any(tag.startswith(value) for tag in self._tags)
+        return tag in self._tags
 
-    def endswith(self, value):
-        """ Implement an endswith method that returns true if any of the tags endswith value
-        """
-        return any(tag.endswith(value) for tag in self._tags)
+    def get_subtags(self, tags=None):
+        """ Return a new TagGroup instance with all the subtags. """
+        return TagGroup(set(self._yield_subtags(tags or self._tags)))
 
-    def contains(self, value):
-        """ Implement a contains method that returns true if any of the tags contains value
-        """
-        return any(value in tag for tag in self._tags)
+    def startswith(self, value, consider_separate_subtags=False, tags=None):
```

Comment:
Does python also provide an alias of `startswith` that is `starts_with`?

Reply:
On further reflection, I think we should go with snake case because that's what it should be and (hopefully) will eventually be updated to in (maybe) python4

Reply:
if in doubt go with what matches the style guide! `starts_with` it is. Hopefully we'll get some Scientists who are learning to code using the twined framework; in that event we don't want to confuse them with historical legacy.
If you really wanted to be smart you could alias the `startswith` command with a deprecation warning but I think that's overkill since we never had it in the first place.

We shouldn't try to anticipate what'll be in Python4 though, unless there's a draft spec somewhere... (is it even being worked on yet?!)